File: online_store/domain/models/Customer.py
from online_store.domain.models.ShoppingCart import ShoppingCart
from online_store.domain.models.Catalog import Catalog
from online_store.domain.models.products.ProductInterface import ProductInterface
from abc import ABC, abstractmethod

class CustomerInterface(ABC):
    @abstractmethod
    def check_item_in_catalog(self, catalog, title):
        pass

    @abstractmethod
    def add_item_to_cart(self, cart, item):
        pass

class AdminInterface(ABC):
    @abstractmethod
    def add_item_to_catalog(self, catalog, item: ProductInterface):
        pass


class Customer(CustomerInterface):
    def __init__(self, name: str, email: str):
        self.name = name
        self.email = email
        self.shopping_cart = ShoppingCart()

    def check_item_in_catalog(self, catalog: Catalog, item):
        return catalog.search_item(item) is not None

    def add_item_to_cart(self, catalog: Catalog, item):
        if self.check_item_in_catalog(catalog, item):
            self.shopping_cart.add_item(item)
            catalog.remove_item(item)
            return f"Item added"
        return f"Item does not exist"
    
    def clear_cart(self):
        self.shopping_cart.items = []

class Admin(AdminInterface):
    def __init__(self, name: str, email: str):
        self.name = name
        self.email = email

    def add_item_to_catalog(self, catalog:Catalog, item: ProductInterface):
        catalog.add_item(item)

File: tmpsjava/online_shop/src/domain/models/Customer.java
package domain.models;

import domain.models.products.ProductBundle;
import domain.models.products.ProductInterface;

import java.util.List;

public class Customer implements CustomerInterface{
    private String name;
    private String email;
    public ShoppingCartInterface shopping_cart;

    public Customer(String name, String email, ShoppingCartInterface sc){
        this.name = name;
        this.email = email;
        this.shopping_cart = sc;
    }
    @Override
    public boolean checkItemInCatalog(Catalog catalog, ProductInterface item) {
        return catalog.search_item(item) != null;
    }

    @Override
    public void addItemToCart(Catalog catalog, ProductInterface item) {
        if(this.checkItemInCatalog(catalog, item)){
            this.shopping_cart.add_item(item);
            System.out.println(this.name + ":" + this.email + " has added in his/hers shopping cart " +
                    item.getTitle() + ":" + System.identityHashCode(item));
            return;
        }else if (item instanceof ProductBundle){
            System.out.println(System.identityHashCode(item) + ":" + item.getTitle() + " is being added to shopping cart");
            ProductBundle bundle = (ProductBundle) item;
            List<ProductInterface> bundleItems = bundle.getItems();
            for(ProductInterface bit : bundleItems){
                this.addItemToCart(catalog, bit);
            }
            return;
        }
        System.out.println(System.identityHashCode(item) + ":" + item.getTitle() + " is not in the catalog:"+System.identityHashCode(catalog));
    }

    @Override
    public ShoppingCartInterface getShoppingCart(){
        return this.shopping_cart;
    }
}
